README.md
# duration - Time duration utilities

_Formerly part of [es5-ext](https://github.com/medikoo/es5-ext) project._

## Installation

### Node.js

	$ npm install duration

### Browser

Can be bundled for browser with help of [modules-webmake](https://github.com/medikoo/modules-webmake)

## Example usage:

```javascript
var Duration = require('duration')

var duration = new Duration(new Date(2000, 6, 7),
	new Date(2010, 8, 13, 3, 23, 8, 456));

console.log("Years: ", duration.years);
console.log("Months: ", duration.months);
console.log("Days: ", duration.days);
console.log("Hours: ", duration.hours);
console.log("Minutes: ", duration.minutes);
console.log("Seconds: ", duration.seconds);
console.log("Milliseconds: ", duration.milliseconds);

console.log("Trailing months: ", duration.month);
console.log("Trailing days: ", duration.day);
console.log("Trailing hours: ", duration.hour);
console.log("Trailing minutes: ", duration.minute);
console.log("Trailing seconds: ", duration.second);
console.log("Trailing milliseconds: ", duration.millisecond);

console.log("Default string representation: ", duration.toString());
console.log("Alternative string representation: ", duration.toString(1));
console.log("Custom string representation: ",
	duration.toString("H: %Hs m: %M"));
```

Output:

```
Years:  10
Months:  122
Days:  3720
Hours:  89283
Minutes:  5357003
Seconds:  321420188
Milliseconds:  321420188456
Trailing months:  2
Trailing days:  6
Trailing hours:  3
Trailing minutes:  23
Trailing seconds:  8
Trailing milliseconds:  456
Default string representation:  10y 2m 6d 03:23:08.456
Alternative string representation:  10y 2m 6d 3h 23m 8s 456ms
Custom string representation:  H: 89283 m: 23
```

## Duration(from[, to])

Main module is both constructor and factory method, and can be used either way.  
`from` and `to` are expected to be JavaScript Date objects. `to` is optional, and if not provided it defaults to current time.

## Duration.prototype properties

### years

Returns full years of the duration

### months

Returns full months of the duration

### days

Returns full days of the duration

### hours

Returns full hours of the duration

### seconds

Returns full seconds of the duration

### minutes

Returns full minutes of the duration

### milliseconds

Returns milliseconds of the duration

### year

Same as `years`. Returns full years of the duration

### month

Returns trailing months of the duration

### day

Returns trailing days of the duration

### hour

Returns trailing hours of the duration

### minute

Returns trailing minutes of the duration

### second

Returns trailing seconds of the duration

### millisecond

Returns trailing seconds of the duration

## valueOf()

Same as `milliseconds`. Returns milliseconds of the duration

## toString([mode[, threshold]])
## toString(format)

Returns readable representation of the duration.  
When invoked without arguments (defaults to _mode=0_), returns as:

	10y 2m 6d 03:23:08.456

When invoked with mode `1`, returns alternative representation:

	10y 2m 6d 3h 23m 8s 456ms

Representation returned by default modes can be customized with threshold setting that trims lowest units:

```javascript
duration.toString();     // 10y 2m 6d 03:23:08.456
duration.toString(0, 1); // 10y 2m 6d 03:23:08
duration.toString(0, 2); // 10y 2m 6d 03:23

duration.toString(1);    // 10y 2m 6d 3h 23m 8s 456ms
duration.toString(1, 1); // 10y 2m 6d 3h 23m 8s
duration.toString(1, 2); // 10y 2m 6d 3h 23m
```

When invoked with string, formats the duration according to given string format, where:

* `%y` - `duration.year`
* `%m` - `duration.month`
* `%d` - `duration.day`
* `%H` - `duration.hour`
* `%M` - `duration.minute`
* `%S` - `duration.second`
* `%L` - `duration.millisecond`
* `%ms` - `duration.months`
* `%ds` - `duration.days`
* `%Hs` - `duration.hours`
* `%Ms` - `duration.minutes`
* `%Ss` - `duration.seconds`
* `%Ls` - `duration.milliseconds`
* `%sign` - If duration is negative outputs `-` otherwise empty string

## Tests [![Build Status](https://secure.travis-ci.org/medikoo/duration.png?branch=master)](https://secure.travis-ci.org/medikoo/es5-ext)

	$ npm test


CHANGES
v0.1.3  --  2013.01.02
* Fix documentation of npm install command

v0.1.2  --  2012.11.10
* Fix handling of negative durations (when to is older than from)

v0.1.1  --  2012.10.04
Maintanance
* Rename lib/index.js to lib/duration.js
* Lint cleanup
* Upgrade es5-ext

v0.1.0  --  2012.05.28
* Initial version (derived from es5-ext project)


test/index.js
'use strict';

var copy = require('es5-ext/lib/Date/prototype/copy');

module.exports = function (t) {
	var d1 = new Date(Date.UTC(2001, 1, 2, 1, 1, 1, 1)), d, d2;

	return {
		"Second date is optional": function (a) {
			var d3, m;
			d2 = new Date();
			d = t(d1);
			d3 = new Date();
			a.ok(((m = d.milliseconds) >= t(d1, d2).milliseconds) &&
				(m <= t(d1, d3).milliseconds));
		},
		"Milliseconds": function (a) {
			// 11 milliseconds
			d2 = copy.call(d1);
			d2.setMilliseconds(d2.getMilliseconds() + 11);
			d = t(d1, d2);

			a(d.milliseconds, 11, "Milliseconds");
			a(d.valueOf(), d.milliseconds, "Value");
			a(d.seconds, 0, "Seconds");
			a(d.minutes, 0, "Minutes");
			a(d.hours, 0, "Hours");
			a(d.days, 0, "Days");
			a(d.months, 0, "Months");
			a(d.years, 0, "Years");

			a(d.millisecond, 11, "Trailing milliseconds");
			a(d.second, 0, "Trailing seconds");
			a(d.minute, 0, "Trailing minutes");
			a(d.hour, 0, "Trailing hours");
			a(d.day, 0, "Trailing days");
			a(d.month, 0, "Trailing months");
			a(d.year, 0, "Trailing years");

			a(d.toString(), '.011', 'String presentation');
			a(d.toString(1), '11ms', 'String presentation #2');
			a(d.toString(' %L.%Ls.%S.%Ss.%M.%Ms.%H.%Hs.%d.%ds.%m.%ms.%y '),
				' 011.11.00.0.00.0.00.0.00.0.00.0.0 ', 'String presentation (custom)');
		},
		"Seconds": function (a) {
			// 7 seconds 123 milliseconds
			d2 = copy.call(d1);
			d2.setMilliseconds(d2.getMilliseconds() + 123);
			d2.setSeconds(d2.getSeconds() + 7);
			d = t(d1, d2);

			a(d.milliseconds, 7 * 1000 + 123, "Milliseconds");
			a(d.valueOf(), d.milliseconds, "Value");
			a(d.seconds, 7, "Seconds");
			a(d.minutes, 0, "Minutes");
			a(d.hours, 0, "Hours");
			a(d.days, 0, "Days");
			a(d.months, 0, "Months");
			a(d.years, 0, "Years");

			a(d.millisecond, 123, "Trailing milliseconds");
			a(d.second, 7, "Trailing seconds");
			a(d.minute, 0, "Trailing minutes");
			a(d.hour, 0, "Trailing hours");
			a(d.day, 0, "Trailing days");
			a(d.month, 0, "Trailing months");
			a(d.year, 0, "Trailing years");

			a(d.toString(), '7.123', 'String presentation');
			a(d.toString(1), '7s 123ms', 'String presentation #2');
			a(d.toString(' %L.%Ls.%S.%Ss.%M.%Ms.%H.%Hs.%d.%ds.%m.%ms.%y '),
				' 123.' + (7 * 1000 + 123) + '.07.7.00.0.00.0.00.0.00.0.0 ',
				'String presentation (custom)');
		},
		"Minutes": function (a) {
			// 7 minutes 12 seconds 123 milliseconds
			d2 = copy.call(d1);
			d2.setMilliseconds(d2.getMilliseconds() + 123);
			d2.setSeconds(d2.getSeconds() + 12);
			d2.setMinutes(d2.getMinutes() + 7);
			d = t(d1, d2);

			a(d.milliseconds, 7 * 60 * 1000 + 12 * 1000 + 123, "Milliseconds");
			a(d.valueOf(), d.milliseconds, "Value");
			a(d.seconds, 7 * 60 + 12, "Seconds");
			a(d.minutes, 7, "Minutes");
			a(d.hours, 0, "Hours");
			a(d.days, 0, "Days");
			a(d.months, 0, "Months");
			a(d.years, 0, "Years");

			a(d.millisecond, 123, "Trailing milliseconds");
			a(d.second, 12, "Trailing seconds");
			a(d.minute, 7, "Trailing minutes");
			a(d.hour, 0, "Trailing hours");
			a(d.day, 0, "Trailing days");
			a(d.month, 0, "Trailing months");
			a(d.year, 0, "Trailing years");

			a(d.toString(), '07:12.123', 'String presentation');
			a(d.toString(1), '7m 12s 123ms', 'String presentation #2');
			a(d.toString(' %L.%Ls.%S.%Ss.%M.%Ms.%H.%Hs.%d.%ds.%m.%ms.%y '),
				' 123.' + (7 * 60 * 1000 + 12 * 1000 + 123) + '.12.' + (7 * 60 + 12) +
				'.07.7.00.0.00.0.00.0.0 ', 'String presentation (custom)');
		},
		"Hours": function (a) {
			// 4 hours 7 minutes 12 seconds 123 milliseconds
			d2 = copy.call(d1);
			d2.setMilliseconds(d2.getMilliseconds() + 123);
			d2.setSeconds(d2.getSeconds() + 12);
			d2.setMinutes(d2.getMinutes() + 7);
			d2.setHours(d2.getHours() + 4);
			d = t(d1, d2);

			a(d.milliseconds,
				4 * 60 * 60 * 1000 + 7 * 60 * 1000 + 12 * 1000 + 123, "Milliseconds");
			a(d.valueOf(), d.milliseconds, "Value");
			a(d.seconds, 4 * 60 * 60 + 7 * 60 + 12, "Seconds");
			a(d.minutes, 4 * 60 + 7, "Minutes");
			a(d.hours, 4, "Hours");
			a(d.days, 0, "Days");
			a(d.months, 0, "Months");
			a(d.years, 0, "Years");

			a(d.millisecond, 123, "Trailing milliseconds");
			a(d.second, 12, "Trailing seconds");
			a(d.minute, 7, "Trailing minutes");
			a(d.hour, 4, "Trailing hours");
			a(d.day, 0, "Trailing days");
			a(d.month, 0, "Trailing months");
			a(d.year, 0, "Trailing years");

			a(d.toString(), '04:07:12.123', 'String presentation');
			a(d.toString(1), '4h 7m 12s 123ms', 'String presentation #2');
			a(d.toString(' %L.%Ls.%S.%Ss.%M.%Ms.%H.%Hs.%d.%ds.%m.%ms.%y '),
				' 123.' + (4 * 60 * 60 * 1000 + 7 * 60 * 1000 + 12 * 1000 + 123) +
				'.12.' + (4 * 60 * 60 + 7 * 60 + 12) + '.07.' + (4 * 60 + 7) +
				'.04.4.00.0.00.0.0 ', 'String presentation (custom)');
		},
		"Days": function (a) {
			// 2 days 14 hours 7 minutes 12 seconds 123 milliseconds
			d2 = copy.call(d1);
			d2.setMilliseconds(d2.getMilliseconds() + 123);
			d2.setSeconds(d2.getSeconds() + 12);
			d2.setMinutes(d2.getMinutes() + 7);
			d2.setHours(d2.getHours() + 14);
			d2.setDate(d2.getDate() + 2);
			d = t(d1, d2);

			a(d.milliseconds, 2 * 24 * 60 * 60 * 1000 + 14 * 60 * 60 * 1000 +
				7 * 60 * 1000 + 12 * 1000 + 123, "Milliseconds");
			a(d.valueOf(), d.milliseconds, "Value");
			a(d.seconds, 2 * 24 * 60 * 60 + 14 * 60 * 60 + 7 * 60 + 12,
				"Seconds");
			a(d.minutes, 2 * 24 * 60 + 14 * 60 + 7, "Minutes");
			a(d.hours, 2 * 24 + 14, "Hours");
			a(d.days, 2, "Days");
			a(d.months, 0, "Months");
			a(d.years, 0, "Years");

			a(d.millisecond, 123, "Trailing milliseconds");
			a(d.second, 12, "Trailing seconds");
			a(d.minute, 7, "Trailing minutes");
			a(d.hour, 14, "Trailing hours");
			a(d.day, 2, "Trailing days");
			a(d.month, 0, "Trailing months");
			a(d.year, 0, "Trailing years");

			a(d.toString(), '2d 14:07:12.123', 'String presentation');
			a(d.toString(1), '2d 14h 7m 12s 123ms', 'String presentation #2');
			a(d.toString(' %L.%Ls.%S.%Ss.%M.%Ms.%H.%Hs.%d.%ds.%m.%ms.%y '),
				' 123.' + (2 * 24 * 60 * 60 * 1000 + 14 * 60 * 60 * 1000 +
					7 * 60 * 1000 + 12 * 1000 + 123) + '.12.' +
				(2 * 24 * 60 * 60 + 14 * 60 * 60 + 7 * 60 + 12) + '.07.' +
				(2 * 24 * 60 + 14 * 60 + 7) + '.14.' + (2 * 24 + 14) +
				'.02.2.00.0.0 ', 'String presentation (custom)');
		},
		"Large duration": function (a) {
			// Few years, few months
			var days = 365 * 2 + 28 + 31;
			d2 = new Date(Date.UTC(2003, 3, 2, 1, 1, 1, 1));
			d = t(d1, d2);

			a(d.milliseconds, days * 24 * 60 * 60 * 1000, "Milliseconds");
			a(d.valueOf(), d.milliseconds, "Value");
			a(d.seconds, days * 24 * 60 * 60, "Seconds");
			a(d.minutes, days * 24 * 60, "Minutes");
			a(d.hours, days * 24, "Hours");
			a(d.days, days, "Days");
			a(d.months, 26, "Months");
			a(d.years, 2, "Years");

			a(d.millisecond, 0, "Trailing milliseconds");
			a(d.second, 0, "Trailing seconds");
			a(d.minute, 0, "Trailing minutes");
			a(d.hour, 0, "Trailing hours");
			a(d.day, 0, "Trailing days");
			a(d.month, 2, "Trailing months");
			a(d.year, 2, "Trailing years");

			a(d.toString(), '2y 2m 0d 00:00:00.000', 'String presentation');
			a(d.toString(1), '2y 2m 0d 0h 0m 0s 0ms', 'String presentation #2');
			a(d.toString(' %L.%Ls.%S.%Ss.%M.%Ms.%H.%Hs.%d.%ds.%m.%ms.%y '),
				' 000.' + (days * 24 * 60 * 60 * 1000) + '.00.' +
				(days * 24 * 60 * 60) + '.00.' + (days * 24 * 60) + '.00.' +
				(days * 24) + '.00.' + days + '.02.' + 26 + '.2 ',
				'String presentation (custom)');
		}
	};
};


lib/index.js
'use strict';

var d           = require('es5-ext/lib/Object/descriptor')
  , pad         = require('es5-ext/lib/Number/prototype/pad')
  , date        = require('es5-ext/lib/Date/valid-date')
  , daysInMonth = require('es5-ext/lib/Date/prototype/days-in-month')
  , copy        = require('es5-ext/lib/Date/prototype/copy')
  , dfloor      = require('es5-ext/lib/Date/prototype/floor-day')
  , mfloor      = require('es5-ext/lib/Date/prototype/floor-month')
  , yfloor      = require('es5-ext/lib/Date/prototype/floor-year')
  , format      = require('es5-ext/lib/String/prototype/format')

  , map, valueOf, getYear, Duration;

map = {
	y: function () {
		return String(this.year);
	},
	m: function () {
		return pad.call(this.month, 2);
	},
	d: function () {
		return pad.call(this.day, 2);
	},
	H: function () {
		return pad.call(this.hour, 2);
	},
	M: function () {
		return pad.call(this.minute, 2);
	},
	S: function () {
		return pad.call(this.second, 2);
	},
	L: function () {
		return pad.call(this.millisecond, 3);
	},

	ms: function () {
		return String(this.months);
	},
	ds: function () {
		return String(this.days);
	},
	Hs: function () {
		return String(this.hours);
	},
	Ms: function () {
		return String(this.minutes);
	},
	Ss: function () {
		return String(this.seconds);
	},
	Ls: function () {
		return String(this.milliseconds);
	}
};

Duration = module.exports = function (from, to) {
	// Make it both constructor and factory
	if (!(this instanceof Duration)) {
		return new Duration(from, to);
	}

	this.from = date(from);
	this.to = (to == null) ? new Date() : date(to);
};

Duration.prototype = Object.create(Object.prototype, {
	valueOf: d(valueOf = function () {
		return this.to - this.from;
	}),

	millisecond: d.gs(function () {
		return this.milliseconds % 1000;
	}),
	second: d.gs(function () {
		return this.seconds % 60;
	}),
	minute: d.gs(function () {
		return this.minutes % 60;
	}),
	hour: d.gs(function () {
		return this.hours % 24;
	}),
	day: d.gs(function () {
		var x = copy.call(this.to);
		x.setMonth(x.getMonth() - 1);
		x = daysInMonth.call(x);
		return (x - this.from.getDate() + this.to.getDate()) % x
			- ((this.from - dfloor.call(copy.call(this.from)))
				> (this.to - dfloor.call(copy.call(this.to))));
	}),
	month: d.gs(function () {
		return (12 - this.from.getMonth() + this.to.getMonth()) % 12
			- ((this.from - mfloor.call(copy.call(this.from)))
				> (this.to - mfloor.call(copy.call(this.to))));
	}),
	year: d.gs(getYear = function () {
		return this.to.getFullYear() - this.from.getFullYear()
			- ((this.from - yfloor.call(copy.call(this.from)))
				> (this.to - yfloor.call(copy.call(this.to))));
	}),

	milliseconds: d.gs(valueOf, null),
	seconds: d.gs(function () {
		return Math.floor(this.valueOf() / 1000);
	}),
	minutes: d.gs(function () {
		return Math.floor(this.valueOf() / (1000 * 60));
	}),
	hours: d.gs(function () {
		return Math.floor(this.valueOf() / (1000 * 60 * 60));
	}),
	days: d.gs(function () {
		return Math.floor(this.valueOf() / (1000 * 60 * 60 * 24));
	}),
	months: d.gs(function () {
		return (this.to.getFullYear() - this.from.getFullYear()) * 12
			+ this.to.getMonth() - this.from.getMonth()
			- ((this.from - mfloor.call(copy.call(this.from)))
				> (this.to - mfloor.call(copy.call(this.to))));
	}),
	years: d.gs(getYear),

	toString: d(function (pattern) {
		var s, r;
		if (!pattern) {
			s = "." +  pad.call(this.millisecond, 3);
			if (this.seconds) {
				if (this.minutes) {
					s = pad.call(this.minute, 2) + ":" +
						pad.call(this.second, 2) + s;
					if (this.hours) {
						s = pad.call(this.hour, 2) + ":" + s;
						if (this.days) {
							s = this.day + "d " + s;
							if (this.months) {
								s = this.month + "m " + s;
								if (this.years) {
									s = this.year + "y " + s;
								}
							}
						}
					}
				} else {
					s = this.second + s;
				}
			}
		} else if (pattern === 1) {
			s = this.millisecond + "ms";
			if (this.seconds) {
				s = this.second + "s " + s;
				if (this.minutes) {
					s = this.minute + "m " + s;
					if (this.hours) {
						s = this.hour + "h " + s;
						if (this.days) {
							s = this.day + "d " + s;
							if (this.months) {
								s = this.month + "m " + s;
								if (this.years) {
									s = this.year + "y " + s;
								}
							}
						}
					}
				}
			}
		} else {
			return format.call(pattern, map, this);
		}
		return s;
	})
});


lib/duration.js
'use strict';

var d           = require('es5-ext/lib/Object/descriptor')
  , pad         = require('es5-ext/lib/Number/prototype/pad')
  , date        = require('es5-ext/lib/Date/valid-date')
  , daysInMonth = require('es5-ext/lib/Date/prototype/days-in-month')
  , copy        = require('es5-ext/lib/Date/prototype/copy')
  , dfloor      = require('es5-ext/lib/Date/prototype/floor-day')
  , mfloor      = require('es5-ext/lib/Date/prototype/floor-month')
  , yfloor      = require('es5-ext/lib/Date/prototype/floor-year')
  , toInt       = require('es5-ext/lib/Number/to-int')
  , format      = require('es5-ext/lib/String/prototype/format')

  , abs = Math.abs

  , map, valueOf, getYear, Duration, getCalcData;

map = {
	y: function () { return String(abs(this.year)); },
	m: function () { return pad.call(abs(this.month), 2); },
	d: function () { return pad.call(abs(this.day), 2); },
	H: function () { return pad.call(abs(this.hour), 2); },
	M: function () { return pad.call(abs(this.minute), 2); },
	S: function () { return pad.call(abs(this.second), 2); },
	L: function () { return pad.call(abs(this.millisecond), 3); },

	ms: function () { return String(abs(this.months)); },
	ds: function () { return String(abs(this.days)); },
	Hs: function () { return String(abs(this.hours)); },
	Ms: function () { return String(abs(this.minutes)); },
	Ss: function () { return String(abs(this.seconds)); },
	Ls: function () { return String(abs(this.milliseconds)); },

	sign: function () { return (this.to < this.from) ? '-' : ''; }
};

getCalcData = function (duration) {
	return (duration.to < duration.from) ?
			{ to: duration.from, from: duration.to, sign: -1 } :
			{ to: duration.to, from: duration.from, sign: 1 };
};

Duration = module.exports = function (from, to) {
	// Make it both constructor and factory
	if (!(this instanceof Duration)) return new Duration(from, to);

	this.from = date(from);
	this.to = (to == null) ? new Date() : date(to);
};

Duration.prototype = Object.create(Object.prototype, {
	valueOf: d(valueOf = function () { return this.to - this.from; }),
	millisecond: d.gs(function () { return this.milliseconds % 1000; }),
	second: d.gs(function () { return this.seconds % 60; }),
	minute: d.gs(function () { return this.minutes % 60; }),
	hour: d.gs(function () { return this.hours % 24; }),
	day: d.gs(function () {
		var data = getCalcData(this)
		  , x = copy.call(data.to);
		x.setMonth(x.getMonth() - 1);
		x = daysInMonth.call(x);
		return data.sign * ((x - data.from.getDate() + data.to.getDate()) % x -
			((data.from - dfloor.call(copy.call(data.from))) >
				(data.to - dfloor.call(copy.call(data.to)))));
	}),
	month: d.gs(function () {
		var data = getCalcData(this);
		return data.sign * ((12 - data.from.getMonth() + data.to.getMonth()) % 12 -
			((data.from - mfloor.call(copy.call(data.from))) >
				(data.to - mfloor.call(copy.call(data.to)))));
	}),
	year: d.gs(getYear = function () {
		var data = getCalcData(this);
		return data.sign * (data.to.getFullYear() - data.from.getFullYear() -
			((data.from - yfloor.call(copy.call(data.from))) >
				(data.to - yfloor.call(copy.call(data.to)))));
	}),

	milliseconds: d.gs(valueOf, null),
	seconds: d.gs(function () { return toInt(this.valueOf() / 1000); }),
	minutes: d.gs(function () { return toInt(this.valueOf() / (1000 * 60)); }),
	hours: d.gs(function () { return toInt(this.valueOf() / (1000 * 60 * 60)); }),
	days: d.gs(function () {
		return toInt(this.valueOf() / (1000 * 60 * 60 * 24));
	}),
	months: d.gs(function () {
		var data = getCalcData(this);
		return data.sign * ((data.to.getFullYear() - data.from.getFullYear()) * 12 +
			data.to.getMonth() - data.from.getMonth() -
			((data.from - mfloor.call(copy.call(data.from))) >
				(data.to - mfloor.call(copy.call(data.to)))));
	}),
	years: d.gs(getYear),

	toString: d(function (pattern) {
		var s;
		if (!pattern) {
			s = "." +  pad.call(abs(this.millisecond), 3);
			if (this.seconds) {
				if (this.minutes) {
					s = pad.call(abs(this.minute), 2) + ":" +
						pad.call(abs(this.second), 2) + s;
					if (this.hours) {
						s = pad.call(abs(this.hour), 2) + ":" + s;
						if (this.days) {
							s = abs(this.day) + "d " + s;
							if (this.months) {
								s = abs(this.month) + "m " + s;
								if (this.years) {
									s = abs(this.year) + "y " + s;
								}
							}
						}
					}
				} else {
					s = abs(this.second) + s;
				}
			}
			if (this.to < this.from) (s = '-' + s);
		} else if (pattern === 1) {
			s = abs(this.millisecond) + "ms";
			if (this.seconds) {
				s = abs(this.second) + "s " + s;
				if (this.minutes) {
					s = abs(this.minute) + "m " + s;
					if (this.hours) {
						s = abs(this.hour) + "h " + s;
						if (this.days) {
							s = abs(this.day) + "d " + s;
							if (this.months) {
								s = abs(this.month) + "m " + s;
								if (this.years) {
									s = abs(this.year) + "y " + s;
								}
							}
						}
					}
				}
			}
			if (this.to < this.from) (s = '-' + s);
		} else {
			return format.call(pattern, map, this);
		}
		return s;
	})
});
